### ISSUE TITLE
Refactor code to simplify adding another implementations of libc

### ISSUE THREAD

**Issue by namirem705** [pwndbg/pwndbg#2034]
**Is your feature request related to a problem? Please describe.**
Now it's impossible to use pwndbg with libc other than glibc - musl or uclibc. There are now a lot of glibc/ptmalloc-specific things that makes difficult to support another libc's

**Describe the solution you'd like**
Code refactoring to add more abstractions would be great.

**Comment by disconnect3d** [comment]
@CptGibbon @lebr0nli WDYT? What needs to be done here? How would u see such an abstraction?

**Comment by CptGibbon** [comment]
> @CptGibbon @lebr0nli WDYT? What needs to be done here? How would u see such an abstraction?

I think supporting other standard libraries is a great feature, though it's a lot of work if you want the project to have parity with its own GLIBC tools.

An incremental step towards this goal could be introducing standalone commands that target other standard libraries, then work towards making the code more modular long-term. I know that seems backwards, but it could be easier once we see what those new commands look like.

@namirem705 could you give us an example of a pwndbg command that targets e.g. uclibc that you'd like to see implemented?

**Comment by lebr0nli** [comment]
> @CptGibbon @lebr0nli WDYT? What needs to be done here? How would u see such an abstraction?

> An incremental step towards this goal could be introducing standalone commands that target other standard libraries, then work towards making the code more modular long-term. I know that seems backwards, but it could be easier once we see what those new commands look like.

I agreed with the opinion of @CptGibbon above, that's just what I was thinking :)

**Comment by namirem705** [comment]
@CptGibbon IMO making standalone commands means do same work twice (if it eventually  will be intergated with basic commands), so I prefer to see all heap stuff migrated to independent module.

Some code looks like libc-implementation-agnostic already (e.g. pwndbg.heap.current allocator may be not only DebugSym/Heuristic Heap). Some are not (GLIBC_VERSION var calculated in pwndbg/heap/structs.py with any import is very annoying 😁).
Refactoring will make possible to other ppl to contribute other libc support. Actually I started do it for my own research, but there are enough hardcoded values (and some pain).

**Comment by CptGibbon** [comment]
Hey if you have time to submit a refactoring PR based on your existing work it would be very welcome 👍

**Comment by fidgetingbits** [comment]
I've got some musl mallocng heap stuff ported I'll be sending a PR for soon, so there might be some places I can start adding libc abstractions related to some of that.

Atm I've added new heap commands, and have a glibc heap and mallocng heap category. 

That said I think reusing the same command names for a lot of heap stuff would be a bit weird. For example, mallocng is slab-based, chunks are called slots, an arena is called a context, bins are active groups. If I was working on the musl heap, I wouldn't want to use commands with ptmalloc naming personally. Using command names that reinforce learning the naming of the given algorithms internals helps with analysis imo. 

For other more generic libc commands I think it will make sense to abstract, as well as things like logic for automatically detecting what heap(s) the program is using and then enabling those heaps commands, etc.

**Comment by CptGibbon** [comment]
I'm stoked to have more dynamic memory allocators supported and agree that trying to make them all fit the same mold is probably not going to work well.

Potentially some overlap with @jetchirag's upcoming jemalloc work for GSoC here so perhaps worth communicating with them could help avoid redundant effort.

**Comment by disconnect3d** [comment]
@CptGibbon is this done?

**Comment by CptGibbon** [comment]
> @CptGibbon is this done?

AFAIK no PR was ever submitted to address this.

**Comment by disconnect3d** [comment]
Do we still need to do anything here?

At this point we have: glibc, jemalloc and musl's mallocng support in Pwndbg.

For glibc we even work with different versions and the latest 2.42 which caused some issues (and lacks large tcache or sth support yet). I feel like we can probably close this issue now.

cc: @k4lizen @CptGibbon

**Comment by k4lizen** [comment]
i don't think this should be closed. i think we want a libc abstraction layer which doesn't exist currently.

**Comment by CptGibbon** [comment]
IMO it's a useless issue because it's so vague, no thought went into this.

That said, I'm sure someone else will just replicate it if you close it so keeping it around probably won't do any harm.
